test/nri/nri_test.go

```diff
@@ -0,0 +1,559 @@
+package nri
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	goruntime "runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	pluginSyncTimeout = 5 * time.Second
+	eventTimeout      = 2 * time.Second
+)
+
+func TestPluginRegistration(stdT *testing.T) {
+	var (
+		t = &nriTest{
+			plugins: []*plugin{nil},
+		}
+		expected = []*event{
+			PluginConfigEvent,
+			PluginSyncedEvent,
+		}
+	)
+
+	t.Setup(stdT)
+	t.StartPlugins(!WaitForPluginSync)
+	require.NoError(t, t.plugins[0].VerifyEventStream(expected, true, pluginSyncTimeout),
+		"received plugin registration event stream")
+}
+
+func TestPluginSynchronization(stdT *testing.T) {
+	var (
+		t = nriTest{
+			plugins: []*plugin{nil},
+		}
+		containerCount = 3
+		pods           []string
+		ctrs           []string
+	)
+
+	t.Setup(stdT)
+
+	for i := 0; i < containerCount; i++ {
+		pod, ctr := t.runContainer()
+		pods = append(pods, pod)
+		ctrs = append(ctrs, ctr)
+	}
+
+	synced := t.StartPlugins(WaitForPluginSync)
+	t.VerifyPodIDs(pods, synced[0].pods, "pods synchronized with plugin")
+	t.VerifyContainerIDs(ctrs, synced[0].ctrs, "containers synchronized with plugin")
+}
+
+func TestPodEvents(stdT *testing.T) {
+	t := &nriTest{
+		plugins: []*plugin{nil},
+	}
+
+	t.Setup(stdT)
+	t.StartPlugins(WaitForPluginSync)
+
+	pod := t.createPod()
+	t.stopPod(pod)
+	t.removePod(pod)
+
+	expected := []*event{
+		RunPodEvent(pod),
+		StopPodEvent(pod),
+		RemovePodEvent(pod),
+	}
+	require.NoError(t, t.plugins[0].VerifyEventStream(expected, true, eventTimeout),
+		"received pod event stream")
+}
+
+func TestContainerEvents(stdT *testing.T) {
+	var (
+		timeout = eventTimeout
+		t       = &nriTest{
+			plugins: []*plugin{nil},
+		}
+	)
+
+	t.Setup(stdT)
+	t.StartPlugins(WaitForPluginSync)
+	p := t.plugins[0]
+
+	pod := t.createPod()
+	require.NotNil(t, p.WaitEvent(RunPodEvent(pod), timeout), "pod creation event")
```

Comment:
any oppposition to using ginkgo in these? if we don't, ginkgo won't be able to filter tests out in the future (if we want to run just these or run just another test)

Reply:
This is actually the thing that I'm struggling/fighting with the most. So these are not really a unit tests. It is an end-to-end or integration test client for NRI and a bunch test cases.

The client part implements a rudimentary test infra. It allows test cases toestablish a gRPC CRI connection to cri-o, register one or more NRI plugins with cri-o, make sure images are pulled and present, create pods and containers, and provide some commonly used functions for test result verification.

Then there is a (currently pretty basic) set of NRI test cases implemented. They use the client infra to create/start/stop pods and containers, the NRI interface to register plugins, and both interfaces to verify that the right/expected things happen both in the plugins and to containers.

The golang/testing infra (github.com/stretchr) is merely used as a 'test driver' and for verification. The individual test cases implemented here are then tied in to end-to-end localintegration/bats tests using nri.bats, with each test case implemented here becoming an individual test case there for proper granular test failure/success reporting.

I think these could be rewritten so that they are ginkgo tests. I actually started it like that but was struggling with wrapping my head around what to put where wrt. ginkgo vs. the test infra we need, so in my frustration I switched to plain golang testing.

Now I'm struggling with a few remaining things related to this. One is a few tests failing in some of the integration runs, with something that looks like a timeout (so things just being too slow)... but I'm not sure if that's the real reason. Need to find a way to figure out.

Another one is what you pointed out above: ginkgo does not understand that these are not unit tests so it does not know to ignore them. That's a problem, because the tests cannot be run just like that. They need a properly configured cri-o (which the test cases in nri.bats set up when running them) so they just fail when those are not given during a plain unit test ginkgo run.

As a workaround, I  added and empty ginkgo test suite and a single no-op ginkgo test case and simple logic to TestMain to omit running the tests. This takes care of the ginkgo test failure, but leaves the tests without a coverage report... which still leads to a failure then. I don't know if this could be due to me skipping all tests by not doing a testing.M.Run(), but I suspect it might be. So now I'm thinking I'll try to change this into all tests just doing a testing.T.Skip() instead when we need to omit running.

Anyway, considering all this, can you think of a better way to do any or all of this which would eliminate the problems related to using ginkgo/unit tests for end-to-end testing ?

Reply:
Okay, after trying a few things, I ended up simply making ginkgo skipping the NRI test client.